android-agent/src/main/java/io/opentelemetry/android/internal/services/network/data/Carrier.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.internal.services.network.data
+
+import android.os.Build
+import android.telephony.TelephonyManager
+import androidx.annotation.RequiresApi
+
+/**
+ * This class is internal and not for public use. Its APIs are unstable and can change at any time.
+ */
+@RequiresApi(api = Build.VERSION_CODES.P)
+data class Carrier internal constructor(
+    val id: Int,
+    val name: String?,
+    val mobileCountryCode: String?,
+    val mobileNetworkCode: String?,
+    val isoCountryCode: String?,
+) {
+    constructor(builder: Builder) : this(
+        id = builder.id,
+        name = builder.name,
+        mobileCountryCode = builder.mobileCountryCode,
+        mobileNetworkCode = builder.mobileNetworkCode,
+        isoCountryCode = builder.isoCountryCode,
+    )
+
+    class Builder {
```

Comment:
I think we want to not use builders for data classes. :) I think we should just give the constructor args default values and remove the builder.

Reply:
This is probably a matter of personal taste, but I find long-living builders that get passed around for various piece of code to contribute to the init data to be confusing and potentially error-prone - who knows what has been set when the builder is passed around before being built? At that point, the builder itself is carrying state, which partially defeats the purpose of it generating an immutable object.

That said, I'm sure there are exceptions, though I'd have to think really hard before I'd chose a classic builder over constructors+default args or factory methods.